File: Python/aspirateur.py
class Aspirateur:
    def __init__(self, x_max, y_max, x, y, orientation):
        self.x_max = x_max
        self.y_max = y_max
        self.x = x
        self.y = y
        self.orientation = orientation
    
    def executer_commandes(self, commandes):
        for cmd in commandes:
            if cmd == 'D':
                self.tourner_droite()
            elif cmd == 'G':
                self.tourner_gauche()
            elif cmd == 'A':
                self.avancer()
    
    def tourner_droite(self):
        rotations = {'N': 'E', 'E': 'S', 'S': 'W', 'W': 'N'}
        self.orientation = rotations[self.orientation]
    
    def tourner_gauche(self):
        rotations = {'N': 'W', 'W': 'S', 'S': 'E', 'E': 'N'}
        self.orientation = rotations[self.orientation]
    
    def avancer(self):
        deplacements = {
            'N': (0, 1),
            'E': (1, 0),
            'S': (0, -1),
            'W': (-1, 0)
        }
        dx, dy = deplacements[self.orientation]
        new_x, new_y = self.x + dx, self.y + dy
        
        if 0 <= new_x <= self.x_max and 0 <= new_y <= self.y_max:
            self.x, self.y = new_x, new_y
    
    def position_actuelle(self):
        return (self.x, self.y, self.orientation)

# Test
aspirateur = Aspirateur(10, 10, 5, 5, 'N')
aspirateur.executer_commandes('DADADADAA')
print(aspirateur.position_actuelle())  # (5, 6, 'N')

File: Java/Aspirateur.java
public class Aspirateur {
    private int x;
    private int y;
    private char orientation;
    private final int xMax;
    private final int yMax;

    public Aspirateur(int xMax, int yMax, int x, int y, char orientation) {
        if (x < 0 || x > xMax || y < 0 || y > yMax) {
            throw new IllegalArgumentException("Position initiale invalide");
        }
        if (!"NESW".contains(String.valueOf(orientation))) {
            throw new IllegalArgumentException("Orientation initiale invalide");
        }
        
        this.xMax = xMax;
        this.yMax = yMax;
        this.x = x;
        this.y = y;
        this.orientation = orientation;
    }

    public void executerCommandes(String commandes) {
        for (char cmd : commandes.toCharArray()) {
            switch (cmd) {
                case 'D': tournerDroite(); break;
                case 'G': tournerGauche(); break;
                case 'A': avancer(); break;
                default: 
                    throw new IllegalArgumentException("Commande invalide: " + cmd);
            }
        }
    }

    private void tournerDroite() {
        switch (orientation) {
            case 'N': orientation = 'E'; break;
            case 'E': orientation = 'S'; break;
            case 'S': orientation = 'W'; break;
            case 'W': orientation = 'N'; break;
        }
    }

    private void tournerGauche() {
        switch (orientation) {
            case 'N': orientation = 'W'; break;
            case 'W': orientation = 'S'; break;
            case 'S': orientation = 'E'; break;
            case 'E': orientation = 'N'; break;
        }
    }

    private void avancer() {
        int newX = x, newY = y;
        switch (orientation) {
            case 'N': newY++; break;
            case 'E': newX++; break;
            case 'S': newY--; break;
            case 'W': newX--; break;
        }

        if (newX >= 0 && newX <= xMax && newY >= 0 && newY <= yMax) {
            x = newX;
            y = newY;
        }
    }

    public String positionActuelle() {
        return String.format("(%d, %d, %c)", x, y, orientation);
    }

    public static void main(String[] args) {
        try {
            Aspirateur aspirateur = new Aspirateur(10, 10, 5, 5, 'N');
            aspirateur.executerCommandes("DADADADAA");
            System.out.println(aspirateur.positionActuelle()); // (5, 6, N)
        } catch (Exception e) {
            System.err.println("Erreur: " + e.getMessage());
        }
    }
}